docs/randomized_testing.md

```diff
@@ -168,6 +168,609 @@ roughly the same things from the user (which operations to run, how to generate
+  generators and generated values. Plus, the user sometimes has to typecast
+  these arguments. We saw that in the `CallA` method above.
+- If we support primitive types, there is a bit of inconsistency because the
+  primitive types don't implement the `Generator[T]` interface.
+
+### Option 8
+
+Option 8 is identical to Option 7 with a slight change to the `Generator`
+interface. Option 7 has the oddity that op arguments double as generated values
+and generators. This approach tries to address that. We have two interfaces:
+
+```go
+type Generator[T any] interface {
+    Generate(*rand.Rand) T
+}
+
+type Rand[G Generator[T], T any] interface {
+    Get() T
+}
+```
+
+`Generator[T]` is the same as before. We introduce a `Rand[G, T]` interface to
+represent a random value of type `T` generated by a generator of type `T`. Now,
+op arguments are of type `Rand[G, T]`:
+
+```go
+type gen struct {}
+func (gen) Generate(r *rand.Rand) int {
+    return r.Intn(100)
+}
+
+func (w *Workload) CallA(ctx context.Context, x sim.Rand[gen]) error {
+    if y, err := w.a.Get().A(ctx, x.Get()); err == nil {
+        w.replies[x] = y
+    }
+    return nil
+}
+```
+
+**Pros.**
+
+- There is a separation between generators and generated values.
+- There is no longer a need for typecasting.
+
+**Cons.**
+
+- The API is complicated and a bit confusing.
+
+### Option 9
+
+Option 9 is identical to Option 7 that, like Option 8, tries to avoid values
+doubling as their own generators. With this option, we register generators in
+`Init`.
+
+```go
+type myInt int
+type myIntGen struct{}
+func (myIntGen) Generate(r *rand.Rand) myInt {
+    return r.Randn(100)
+}
+
+func (w *Workload) Init(ctx context.Context, s *sim.Sim) error {
+    // Register myIntGen. Whenever an op takes an argument of type myInt, the
+    // simulator will use myIntGen to generate it.
+    sim.RegisterGenerator[myIntGen](s)
+    return nill
+}
+
+func (w *Workload) CallA(ctx context.Context, x myInt) error {
+    if y, err := w.a.Get().A(ctx, int(x)); err == nil {
+        w.replies[x] = y
+    }
+    return nil
+}
+```
+
+We can simplify this code by using a raw `int` instead of `myInt` as well.
+
+```go
+type intgen struct{}
+func (intgen) Generate(r *rand.Rand) int {
+    return r.Randn(100)
+}
+
+func (w *Workload) Init(ctx context.Context, s *sim.Sim) error {
+    // Register intgen. Whenever an op takes an argument of type int, the
+    // simulator will use intgen to generate it.
+    sim.RegisterGenerator[intgen](s)
+    return nill
+}
+
+func (w *Workload) CallA(ctx context.Context, x int) error {
+    if y, err := w.a.Get().A(ctx, x); err == nil {
+        w.replies[x] = y
+    }
+    return nil
+}
+```
+
+**Pros.**
+
+- There is a separation between generators and generated values.
+- Primitive values can be generated in custom ways.
+
+**Cons.**
+
+- People can easily forget to register a generator for every type.
+- People can register two generators for the same type.
+
+### Decision
```

Comment:
I added the decision we discussed on offline. I'll start implementing things, and we can tweak stuff as we go.